accord-core/src/main/java/accord/utils/SortedArrays.java

```diff
@@ -0,0 +1,535 @@
+                rightIdx++;
+                result[resultSize++] = leftKey;
+            }
+            else
+            {
+                if (cmp < 0) leftIdx++;
+                else rightIdx++;
+            }
+        }
+
+        if (resultSize < result.length)
+            result = Arrays.copyOf(result, resultSize);
+        
+        return result;
+    }
+
+    /**
+     * Given two sorted arrays, return the elements present only in the first, preferentially returning the first array
+     * itself if possible
+     */
+    @SuppressWarnings("unused") // was used until recently, might be used again?
+    public static <T extends Comparable<? super T>> T[] linearDifference(T[] left, T[] right, IntFunction<T[]> allocate)
+    {
+        int rightIdx = 0;
+        int leftIdx = 0;
+
+        T[] result = null;
+        int resultSize = 0;
+
+        while (leftIdx < left.length && rightIdx < right.length)
+        {
+            int cmp = left[leftIdx].compareTo(right[rightIdx]);
+            if (cmp == 0)
+            {
+                resultSize = leftIdx++;
+                ++rightIdx;
+                result = allocate.apply(resultSize + left.length - leftIdx);
+                System.arraycopy(left, 0, result, 0, resultSize);
+                break;
+            }
+            else if (cmp < 0)
+            {
+                ++leftIdx;
+            }
+            else
+            {
+                ++rightIdx;
+            }
+        }
+
+        if (result == null)
+            return left;
+
+        while (leftIdx < left.length && rightIdx < right.length)
+        {
+            int cmp = left[leftIdx].compareTo(right[rightIdx]);
+            if (cmp > 0)
+            {
+                result[resultSize++] = left[leftIdx++];
+            }
+            else if (cmp < 0)
+            {
+                ++rightIdx;
+            }
+            else
+            {
+                ++leftIdx;
+                ++rightIdx;
+            }
+        }
+
+        if (resultSize < result.length)
+            result = Arrays.copyOf(result, resultSize);
+
+        return result;
+    }
+
+    public static <A, R> A[] sliceWithOverlaps(A[] slice, R[] select, IntFunction<A[]> factory, AsymmetricComparator<A, R> cmp1, AsymmetricComparator<R, A> cmp2)
+    {
+        A[] result;
+        int resultCount;
+        int ai = 0, ri = 0;
+        while (true)
+        {
+            long ari = findNextIntersection(slice, ai, select, ri, cmp1, cmp2, Search.CEIL);
+            if (ari < 0)
+            {
+                if (ai == slice.length)
+                    return slice;
+
+                return Arrays.copyOf(slice, ai);
+            }
+
+            int nextai = (int)(ari >>> 32);
+            if (ai != nextai)
+            {
+                resultCount = ai;
+                result = factory.apply(ai + (slice.length - nextai));
+                System.arraycopy(slice, 0, result, 0, resultCount);
+                ai = nextai;
```

Comment:
why don't we update `ri`?  I am in a debugger trying to see the reason and not seeing a clear reason, setting `ri = (int)ari;` is passing in my tests.

The side effect I see by not setting it is that 

```
            int nextai = exponentialSearch(slice, ai, slice.length, select[ri], cmp2, Search.FLOOR) + 1;
             while (ai < nextai)
                 result[resultCount++] = slice[ai++];
```

is always a wasted operation on the first call, this is because `ri` wasn't updated so we *know* a new `ri` is the next match so we search for a match that doesn't exist.

Reply:
I added it in cd51a71795d48010be07aab7c583ac9bfd19913f with docs for this method

Reply:
Yes, I think you're right. I'll think about it more carefully with fresh eyes in the morning.